crates/bevy_ecs/src/error/bevy_error.rs

```diff
@@ -0,0 +1,208 @@
+use alloc::boxed::Box;
+use core::{
+    error::Error,
+    fmt::{Debug, Display},
+};
+
+/// The built in "universal" Bevy error type. This has a blanket [`From`] impl for any type that implements Rust's [`Error`],
+/// meaning it can be used as a "catch all" error.
+///
+/// # Backtraces
+///
+/// When used with the `backtrace` Cargo feature, it will capture a backtrace when the error is constructed (generally in the [`From`] impl]).
+/// When printed, the backtrace will be displayed. By default, the backtrace will be trimmed down to filter out noise. To see the full backtrace,
+/// set the `BEVY_BACKTRACE=full` environment variable.
+///
+/// # Usage
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+///
+/// fn fallible_system() -> Result<(), BevyError> {
+///     // This will result in Rust's built-in ParseIntError, which will automatically
+///     // be converted into a BevyError.
+///     let parsed: usize = "I am not a number".parse()?;
+///     Ok(())
+/// }
+/// ```
+pub struct BevyError {
+    inner: Box<dyn InnerError>,
+}
+
+impl BevyError {
+    /// Creates a new error with the given message.
+    pub fn message<M: Display + Debug + Send + Sync + 'static>(message: M) -> Self {
+        BevyError {
+            inner: Box::new(ErrorImpl {
+                #[cfg(feature = "backtrace")]
+                backtrace: std::backtrace::Backtrace::capture(),
+                error: MessageError(message),
+            }),
+        }
+    }
+
+    /// Attempts to downcast the internal error to the given type.
+    pub fn downcast_ref<E: Error + 'static>(&self) -> Option<&E> {
+        self.inner.error().downcast_ref::<E>()
+    }
+}
+
+trait InnerError: Send + Sync + 'static {
+    #[cfg(feature = "backtrace")]
+    fn backtrace(&self) -> &std::backtrace::Backtrace;
+    fn error(&self) -> &(dyn Error + Send + Sync + 'static);
+}
+
+struct ErrorImpl<E: Error + Send + Sync + 'static> {
+    error: E,
+    #[cfg(feature = "backtrace")]
+    backtrace: std::backtrace::Backtrace,
+}
+
+impl<E: Error + Send + Sync + 'static> InnerError for ErrorImpl<E> {
+    #[cfg(feature = "backtrace")]
+    fn backtrace(&self) -> &std::backtrace::Backtrace {
+        &self.backtrace
+    }
+
+    fn error(&self) -> &(dyn Error + Send + Sync + 'static) {
+        &self.error
+    }
+}
+
+impl<E> From<E> for BevyError
+where
+    E: Error + Send + Sync + 'static,
+{
+    #[cold]
+    fn from(error: E) -> Self {
+        BevyError {
+            inner: Box::new(ErrorImpl {
+                error,
+                #[cfg(feature = "backtrace")]
+                backtrace: std::backtrace::Backtrace::capture(),
+            }),
+        }
+    }
+}
+
+impl Display for BevyError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        writeln!(f, "{}", self.inner.error())?;
+        Ok(())
+    }
+}
+
+impl Debug for BevyError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
```

Comment:
could we add a test for the string filtering? It feels like it'd be pretty fragile to code changes and rust changes.

Reply:
Writing this test in a way that is actually useful involves setting the RUST_BACKTRACE env variable from the test, which is unsafe to do as it could run in parallel. I'm cool with just accepting that though an